File: gacyber_toolkit/02-Scanning/Nmap/Port_Scanning.pl
#!/usr/bin/env perl
use strict;
use warnings;

# Function to perform a port scan using nmap and save results
sub port_scan_and_save {
    my ($ip) = @_;

    # Run nmap to perform a port scan on the host
    print "[*] Scanning $ip for open ports...\n";
    my $scan_result = `nmap -sV $ip`;

    # Save the scan results to a file named after the host's IP address
    my $filename = "$ip.txt";
    open my $fh, '>', $filename or die "Could not open file '$filename' for writing: $!";
    print $fh $scan_result;
    close $fh;

    print "[+] Scan results saved to $filename\n";
}

# Function to discover live hosts in the subnet
sub discover_live_hosts {
    my ($subnet) = @_;
    my @live_hosts;

    print "[*] Discovering live hosts in subnet: $subnet\n";

    # Use nmap to perform a ping sweep of the subnet
    my $ping_sweep_result = `nmap -sn $subnet`;
    @live_hosts = $ping_sweep_result =~ /Nmap scan report for (\d+\.\d+\.\d+\.\d+)/g;

    if (@live_hosts) {
        print "[+] Found live hosts:\n";
        print "    $_\n" for @live_hosts;
    } else {
        print "[-] No live hosts found in the subnet.\n";
    }

    return @live_hosts;
}

# Main script
print "Enter the target subnet (e.g., 192.168.1.0/24): ";
my $subnet = <STDIN>;
chomp($subnet);

# Discover live hosts in the subnet
my @live_hosts = discover_live_hosts($subnet);

# Perform port scans on each live host and save results
if (@live_hosts) {
    for my $host (@live_hosts) {
        port_scan_and_save($host);
    }
} else {
    print "No live hosts to scan.\n";
}

File: gacyber_toolkit/CheatSheets/Port_Scanning.py
import nmap
import os

def scan_subnet(subnet_range):
    """
    Perform a port scan on a target subnet and save the results for each live host in a separate file, including services.

    Args:
        subnet_range (str): The subnet to scan (e.g., "192.168.1.0/24").
    """
    # Initialize the Nmap scanner
    nm = nmap.PortScanner()

    print(f"Scanning subnet: {subnet_range}")
    
    # Scan the subnet for live hosts
    nm.scan(hosts=subnet_range, arguments='-sn')
    live_hosts = [host for host in nm.all_hosts() if nm[host].state() == 'up']

    print(f"Found {len(live_hosts)} live hosts.")
    
    # Perform port scans on each live host
    for host in live_hosts:
        print(f"Scanning ports for host: {host}")
        nm.scan(host, arguments='-sS -sV')  # -sV for service version detection

        # Save the results in a file named after the host's IP address
        file_name = f"{host}.txt"
        with open(file_name, 'w') as file:
            file.write(f"Scan results for {host}:\n")
            file.write("-" * 40 + "\n")
            for proto in nm[host].all_protocols():
                ports = nm[host][proto].keys()
                for port in sorted(ports):
                    state = nm[host][proto][port]['state']
                    service = nm[host][proto][port].get('name', 'unknown')
                    file.write(f"Port {port}/{proto} is {state} (Service: {service})\n")
        
        print(f"Results saved to {file_name}")

if __name__ == "__main__":
    # Specify the subnet range to scan
    subnet_range = input("Enter the subnet range (e.g., 192.168.1.0/24): ")
    scan_subnet(subnet_range)